### PR TITLE
Fix issue 473

### PR DESCRIPTION
**Pull request by boaks**
The first commit uses the stripped executor to fix the race condition.
The seconds removes also the not longer required send worker.
Fix issue #473

Signed-off-by: Achim Kraus <achim.kraus@bosch-si.com>

### CODE REVIEW

File: scandium-core/src/main/java/org/eclipse/californium/scandium/DTLSConnector.java [hunk 1]
```diff
@@ -1218,26 +1206,37 @@ public final void send(final RawData msg) {
...
-
-				@Override
-				public Object getStripe() {
-					return msg.getEndpointContext().getPeerAddress();
-				}
-
-				@Override
-				public void run() {
-					if (running.get()) {
+			if (0 <= pendingOutboundMessages.decrementAndGet()) {
```
Comment (sbernard31) [inline scandium-core/src/main/java/org/eclipse/californium/scandium/DTLSConnector.java]: MISRA ghost come back !

File: scandium-core/src/main/java/org/eclipse/californium/scandium/DTLSConnector.java [hunk 2]
```diff
@@ -1412,66 +1414,78 @@ private void sendFlight(DTLSFlight flight) {
...
+				throw new IOException("Socket closed.");
 			}
-		} else {
-			LOGGER.debug("Socket [{}] is closed, discarding packet ...", config.getAddress());
-			throw new IOException("Socket closed.");
+		} catch (InterruptedException e) {
+			// this means that the worker thread for sending
+			// outbound messages has been interrupted, most
+			// probably because the connector is shutting down
+			Thread.currentThread().interrupt();
```
Comment (sbernard31) [inline scandium-core/src/main/java/org/eclipse/californium/scandium/DTLSConnector.java]: I'm not sure to understand the purpose of this ?

**Comment by boaks** [comment]
Release 1.0.6:

```
protected DTLSFlight receivedHelloVerifyRequest(HelloVerifyRequest message) throws HandshakeException {
   clientHello.setCookie(message.getCookie()); // line 343

```

I suspect the `clientHello` to be `null`. That's set by `getStartHandshakeMessage`.
So receiving "Hello Verify" during execution of `sendNextMessageOverNetwork`, after the handshaker was created, but before the `getStartHandshakeMessage` was called may have caused the `NullPointerException`. 

To fix that, I use the striped executor for the "sending" as well.
And doing so, I got aware, that the `sender` Worker and the `outboundMessages` seems to be just wasting resources. Therefore I added the second commit to remove that stuff also.

**Comment by sbernard31** [comment]
>So receiving "Hello Verify" during execution of sendNextMessageOverNetwork, after the handshaker was created, but before the getStartHandshakeMessage was called may have caused the NullPointerException.

It make senses.
And in general way I think its a good thing that there is no concurrent call for a same peer.

About worker, for the reason I expose above, I don't know if this is worst or not in term of performance. but it seems obvious this will consume less memory without worker and `BlockingQueue`, so it sounds reasonnable to merge this PR.

(Could you bust the MISRA ghost before :))

**Comment by boaks** [comment]
> Create 500 DTLS example clients, expect to send 500000 messages overall.
> 
> messages per s
> 
> without send worker
> 28538
> 30246
> 30165
> 
> with send worker
> 28431
> 29349
> 30712

So there is no real difference.

**Comment by sbernard31** [comment]
I suspect that performance "issue" could come with large payload and lot of different foreign peers on the same connector.

But I don't think this is worthy go deeper now.

**Pull request merged**